### ISSUE TITLE
AWS_VPC_K8S_CNI_RANDOMIZESNAT environment variable for `spec.networking.amazonvpc` resets to `prng` on every upgrade

### ISSUE THREAD

**Issue by Nyefan** [kubernetes/kops#11144]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

1.19.1

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

1.18.12

**3. What cloud provider are you using?**

aws

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

In `cluster.yaml`

```
spec:
  networking:
    amazonvpc:
      env:
        - name: AWS_VPC_K8S_CNI_EXTERNALSNAT
          value: "false"
        - name: AWS_VPC_K8S_CNI_RANDOMIZESNAT
          value: none
        - name: MAX_ENI
          value: none
        - name: WARM_IP_TARGET
          value: "2"
```

`kops update cluster && kops rolling-update cluster`

**5. What happened after the commands executed?**

From `kubectl get ds -n kube-system aws-node -oyaml`

```
spec:
  template:
    spec:
      containers:
      - env:
        - name: ADDITIONAL_ENI_TAGS
          value: '{}'
        - name: AWS_VPC_CNI_NODE_PORT_SUPPORT
          value: "true"
        - name: AWS_VPC_ENI_MTU
          value: "9001"
        - name: AWS_VPC_K8S_CNI_CONFIGURE_RPFILTER
          value: "false"
        - name: AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG
          value: "false"
        - name: AWS_VPC_K8S_CNI_EXTERNALSNAT
          value: "prng"
        - name: AWS_VPC_K8S_CNI_LOGLEVEL
          value: DEBUG
        - name: AWS_VPC_K8S_CNI_LOG_FILE
          value: /host/var/log/aws-routed-eni/ipamd.log
        - name: AWS_VPC_K8S_CNI_RANDOMIZESNAT
          value: none
        - name: AWS_VPC_K8S_CNI_VETHPREFIX
          value: eni
        - name: AWS_VPC_K8S_PLUGIN_LOG_FILE
          value: /var/log/aws-routed-eni/plugin.log
        - name: AWS_VPC_K8S_PLUGIN_LOG_LEVEL
          value: DEBUG
        - name: DISABLE_INTROSPECTION
          value: "false"
        - name: DISABLE_METRICS
          value: "false"
        - name: ENABLE_POD_ENI
          value: "false"
        - name: MY_NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        - name: WARM_ENI_TARGET
          value: "1"
        - name: CLUSTER_NAME
          value: [REDACTED]
        - name: MAX_ENI
          value: "1"
        - name: WARM_IP_TARGET
          value: "1"
```

**6. What did you expect to happen?**

`AWS_VPC_K8S_CNI_EXTERNALSNAT` should be set to `none`

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2018-10-31T22:42:45Z"
  name: [REDACTED]
spec:
  additionalPolicies:
    node: |
      [
        [REDACTED]
      ]
  api:
    loadBalancer:
      class: Classic
      type: Private
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    environment: [REDACTED]
  cloudProvider: aws
  configBase: s3://[REDACTED]
  dnsZone: [REDACTED]
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
      volumeType: gp3
    - instanceGroup: master-us-east-1b
      name: b
      volumeType: gp3
    - instanceGroup: master-us-east-1c
      name: c
      volumeType: gp3
    name: main
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
      volumeType: gp3
    - instanceGroup: master-us-east-1b
      name: b
      volumeType: gp3
    - instanceGroup: master-us-east-1c
      name: c
      volumeType: gp3
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeControllerManager:
    horizontalPodAutoscalerDownscaleStabilization: 22m0s
  kubeDNS:
    provider: CoreDNS
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.18.12
  masterInternalName: api.internal.[REDACTED]
  masterPublicName: api.[REDACTED]
  networkCIDR: 10.3.0.0/16
  networking:
    amazonvpc:
      env:
      - name: AWS_VPC_K8S_CNI_EXTERNALSNAT
        value: "false"
      - name: AWS_VPC_K8S_CNI_RANDOMIZESNAT
        value: none
      - name: MAX_ENI
        value: "1"
      - name: WARM_IP_TARGET
        value: "1"
      imageName: 602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.6.3
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets: [REDACTED]
  topology:
    bastion:
      bastionPublicName: bastion.[REDACTED]
    dns:
      type: Public
    masters: private
    nodes: private
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**
`AWS_VPC_K8S_CNI_RANDOMIZESNAT` continues to reset to `"prng"` on every upgrade

**Comment by MoShitrit** [comment]
That's probably because the AWS CNI daemonset version gets upgraded between version, which forces the manifest to get re-applied.
In order to work around that- I think the only option is to make that field configurable via the cluster spec.
I can work on that one. 
We would probably keep `prng` as the default value, but can be overridden to any of the possible values taken from the official docs (https://github.com/aws/amazon-vpc-cni-k8s/tree/master#cni-configuration-variables):

> AWS_VPC_K8S_CNI_RANDOMIZESNAT
> 
>   Type: String
>   
>   Default: prng
>   
>   Valid Values: hashrandom, prng, none

However, this also means that we need to start being aware of changes for that config key, in case there are newly-added values or removed values.
@rifelpet @hakman what do you guys think about this? Do we have a current process for such things? Or should I just go ahead and add that and we'll simply need to remember to check that key for changes across upgrades of the CNI?

/assign

**Comment by Nyefan** [comment]
Seeing as the daemonset `spec.template.spec.containers[].env[]` uses the patch merge key `name`, would it make sense to just follow that behavior by merging the `spec.networking.amazonvpc.env[]` array into the `aws-node.yaml` in the correct direction before `kubectl apply`ing it?  It seems that this is already done, but with the `aws-node.yaml` template taking priority where there are collisions rather than the `cluster.yaml`

I don't know what this looks like internally, but the environment variable, `MAX_ENI` is added/updated correctly.  I assume this is because it is not present in [the aws-node template](https://github.com/kubernetes/kops/blob/1b7dd3b5b8460a17fd143c1934f668b1f1a32057/upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.16.yaml.template)

**Comment by MoShitrit** [comment]
> I don't know what's going on internally to kops when managing these addons, so this is all just conjecture, but seeing as the daemonset `spec.template.spec.containers[].env[]` uses the patch merge key `name`, would it make sense to just follow that behavior by merging the `spec.networking.amazonvpc.env[]` array into the `aws-node.yaml` in the correct direction before `kubectl apply`ing it? It seems that this is already done, but with the `aws-node.yaml` template taking priority where there are collisions rather than the `cluster.yaml` because the environment variable, `MAX_ENI` is added/updated correctly. I assume this is because it is not present in [the aws-node template](https://github.com/kubernetes/kops/blob/1b7dd3b5b8460a17fd143c1934f668b1f1a32057/upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.16.yaml.template).

Yeah, I think you're right.
In that case- since this config key (`AWS_VPC_K8S_CNI_RANDOMIZESNAT`) isn't required and already has an assigned value anyway, it _might_ make sense to simply remove it from the template. However- this also means that we would deviate from the [vendor's version](https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/release-1.7/config/v1.7/aws-k8s-cni.yaml) that _does_ include it with the default value.
So the way I see it, we have two options here:
1. Drop [these two lines](https://github.com/kubernetes/kops/blob/50b89a0c253255711f5db523de2755ce4d46da51/upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.16.yaml.template#L125-L126) from the upup template and deviate from the vendor's version.

2. Add `AWS_VPC_K8S_CNI_RANDOMIZESNAT` as a configurable field in the networking spec for aws cni with a default value of `prng`.

I _think_ option 2 would make more sense and align better with our SOP, but I also don't have a strong attachment to neither of options, so I'll wait for the others to chime-in and share their thoughts.

**Comment by hakman** [comment]
@MoShitrit Option 2 makes sense if we would want to make all/most of the defaults configurable.

Option 3 would be to comment all those env vars and create a template function that sets these defaults and overrides them with what amazonvpc.env has added:
https://github.com/kubernetes/kops/blob/97869057129b30ea284c3ed1bdf1db36e752701d/upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.16.yaml.template#L3823-L3824
https://github.com/kubernetes/kops/blob/5e0835d48d5a8c5a2e3208693d5f17448c9d562a/upup/pkg/fi/cloudup/template_functions.go#L148-L161

**Comment by MoShitrit** [comment]
> @MoShitrit Option 2 makes sense if we would want to make all/most of the defaults configurable.
> 
> Option 3 would be to comment all those env vars and create a template function that sets these defaults and overrides them with what amazonvpc.env has added:
> 
> https://github.com/kubernetes/kops/blob/97869057129b30ea284c3ed1bdf1db36e752701d/upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.16.yaml.template#L3823-L3824
> 
> 
> https://github.com/kubernetes/kops/blob/5e0835d48d5a8c5a2e3208693d5f17448c9d562a/upup/pkg/fi/cloudup/template_functions.go#L148-L161

I love option 3! That seems like it's the most flexible to the user.
I'll try to work on that soon.

**Comment by MoShitrit** [comment]
@hakman 
I finally got around to look at that issue a bit deeper.
I found an oversight from my previous list of possibilities, which technically lines up pretty well with what you suggested, without having to manually specify all the possible config variables. 
`AmazonVpcNetworkingSpec` already supports an array of environment variables ([see here](https://github.com/kubernetes/kops/blob/5582e27adb3e768a3a0315c9e4dfd531a68d4797/pkg/apis/kops/networking.go#L256)), so essentially, all we have to do it remove all the pre-assigned variables from the template ([these](https://github.com/kubernetes/kops/blob/640bcc9d036730075248c373cf47aa419f2b09b1/upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.16.yaml.template#L109-L138)).
I've looked at the documentation for aws-cni and [confirmed that these are all optional and pre-assigned with their default values](https://github.com/aws/amazon-vpc-cni-k8s#cni-configuration-variables).

So my idea is to simply remove all the hardcoded variables from the template, and then add a link in the documentation pointing [here](https://github.com/aws/amazon-vpc-cni-k8s#cni-configuration-variables) and instructing users to simply add whichever config overrides they need under `env` section.
This would guarantee the maximum flexibility with the least amount of effort.
I'd also love to hear @rifelpet 's thoughts on that one 😄 
WDYT?
